protocol/vm/vm.go

```diff
@@ -55,39 +59,43 @@ func verifyTxInput(tx *bc.Tx, inputIndex int) (bool, error) {
 
 	txinput := tx.Inputs[inputIndex]
 
-	var program []byte
-	switch inp := txinput.TypedInput.(type) {
-	case *bc.IssuanceInput:
-		if inp.VMVersion != 1 {
-			return false, ErrUnsupportedVM
-		}
-		program = inp.IssuanceProgram
-	case *bc.SpendInput:
-		if inp.VMVersion != 1 {
+	expansionReserved := tx.Version == 1
+
+	sigHasher := bc.NewSigHasher(&tx.TxData)
+
+	f := func(vmversion uint64, prog []byte, args [][]byte) (bool, error) {
+		if vmversion != 1 {
 			return false, ErrUnsupportedVM
 		}
-		program = inp.ControlProgram
-	default:
-		return false, ErrUnsupportedTx
-	}
 
-	vm := virtualMachine{
-		tx:         tx,
-		inputIndex: inputIndex,
-		sigHasher:  bc.NewSigHasher(&tx.TxData),
+		vm := virtualMachine{
+			tx:         tx,
+			inputIndex: inputIndex,
+			sigHasher:  sigHasher,
 
-		program:  program,
-		runLimit: initialRunLimit,
-	}
+			expansionReserved: expansionReserved,
 
-	for _, arg := range txinput.Arguments() {
-		err := vm.push(arg, false)
-		if err != nil {
-			return false, err
+			mainprog: prog,
+			curprog:  prog,
+			runLimit: initialRunLimit,
+		}
+		for _, arg := range args {
+			err := vm.push(arg, false)
+			if err != nil {
+				return false, err
+			}
 		}
+		ok, err := vm.run()
+		return ok, wrapErr(err, &vm, args)
 	}
 
-	return vm.run()
+	switch inp := txinput.TypedInput.(type) {
```

Comment:
I'm still a little surprised we don't yet have a method `TxInput.Program`. Places like this (which is most of them) should just call that method. Stay tuned for a patch.